cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -1520,28 +1523,44 @@ private Map<ProgramRunId, RunRecordDetail> getUnfinishedRuns(Set<ProgramRunId> p
-                                               RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS)));
+                                               RunIds.getTime(programRunId.getRun(), TimeUnit.SECONDS),
+                                               false));
     }
-    return getRunRecordsTable().multiRead(allKeys).stream()
-      .map(AppMetadataStore::deserializeRunRecordMeta)
-      .collect(Collectors.toMap(RunRecordDetail::getProgramRunId, r -> r, (r1, r2) -> {
-        throw new IllegalStateException("Duplicate run record for " + r1.getProgramRunId());
-      }, LinkedHashMap::new));
+    
+    return getRunsByKeys(allKeys);
   }
 
   private Map<ProgramRunId, RunRecordDetail> getCompletedRuns(Set<ProgramRunId> programRunIds) throws IOException {
     List<List<Field<?>>> allKeys = new ArrayList<>();
     for (ProgramRunId programRunId : programRunIds) {
-      List<Field<?>> keys = getRunRecordProgramPrefix(TYPE_RUN_RECORD_COMPLETED, programRunId.getParent());
+      // Get all keys without version
+      List<Field<?>> keysWithoutVersion = getRunRecordProgramPrefix(TYPE_RUN_RECORD_COMPLETED,
+                                                                    programRunId.getParent(),
+                                                                    false);
       // Get start time from RunId
       long programStartSecs = RunIds.getTime(RunIds.fromString(programRunId.getRun()), TimeUnit.SECONDS);
-      keys.add(Fields.longField(StoreDefinition.AppMetadataStore.RUN_START_TIME,
-                                getInvertedTsKeyPart(programStartSecs)));
-      keys.add(Fields.stringField(StoreDefinition.AppMetadataStore.RUN_FIELD, programRunId.getRun()));
-      allKeys.add(keys);
+      keysWithoutVersion.add(Fields.longField(StoreDefinition.AppMetadataStore.RUN_START_TIME,
+                                              getInvertedTsKeyPart(programStartSecs)));
+      keysWithoutVersion.add(Fields.stringField(StoreDefinition.AppMetadataStore.RUN_FIELD, programRunId.getRun()));
+      allKeys.add(keysWithoutVersion);
     }
-    return getRunRecordsTable().multiRead(allKeys).stream()
-      .map(AppMetadataStore::deserializeRunRecordMeta)
+
+    return getRunsByKeys(allKeys);
+  }
+
+  private Map<ProgramRunId, RunRecordDetail> getRunsByKeys(List<List<Field<?>>> allKeys) throws IOException {
+    List<RunRecordDetail> runRecordDetails = new ArrayList<>();
+    Collection<Range> ranges = allKeys.stream().map(Range::singleton).collect(Collectors.toList());
+
+    try (CloseableIterator<StructuredRow> iterator =
+           getRunRecordsTable().multiScan(ranges, Integer.MAX_VALUE)) {
+      runRecordDetails =
+        StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, Spliterator.ORDERED), false)
+          .map(AppMetadataStore::deserializeRunRecordMeta)
+          .collect(Collectors.toList());
+    }
+
+    return runRecordDetails.stream()
```

Comment:
Any reason for the intermediate list? Create map right away and return from inside the "try"

Reply:
updated